[WIP][WMS][12.0] Available to Promise Release

guewen opened the pull request:
Implements: #710

Commit:
Add stock_virtual_reservation

Commit:
Add virtual_reserved_qty field

Commit:
Defer creation of chained moves

Commit:
add creation of origin moves on assign (WIP)

Commit:
Add todos

Commit:
Fix backorders, add wizard

Commit:
Move fields around

Commit:
Compute available per warehouse

Commit:
Support "one" delivery type

Commit:
Add sale_stock_virtual_reservation_rule

Glue for stock_virtual_reservation and sale

Commit:
Track moves to release

Commit:
Set decimal precision on field

Commit:
Add test for picking with several moves

Commit:
Fix stock qty and assign

* we want to work with the virtual available as we cannot delivery what
is already outgoing
* call action_assign even if the state is already assigned, because it
may change the quantity

Commit:
Fix different UOM

Commit:
Add readme

florian-dacosta commented on stock_virtual_reservation/models/stock_picking.py:
```diff
@@ -0,0 +1,24 @@
+    def _compute_need_rule_pull(self):
+        for picking in self:
+            picking.need_rule_pull = any(
+                move.need_rule_pull for move in picking
```
should be `move.need_rule_pull for move in picking.move_lines`

florian-dacosta reviewed (commented):
I know it is a wip no worries, just a small comment.
Also, I try to understand well the use of the module, and I wonder about the make to order process.
If you configure one product with make to order and buy route.
On sale order validation, the delivery order will be created and marked as `need_rule_pull`.
So the purchase order for the make to order product won't be created.
Then, the delivery order should be releases once the products are virtually reserved, which won't ever happen for the make to order product, right?

guewen commented:
@florian-dacosta spot on! I'll find a correction for the make to order issue, thanks :)

Commit:
Rename to stock_available_to_promise_release

Commit:
Move release option from warehouse to location route

So we can activate the release only on some routes, while leaving
other routes work as normal (eg. MTO). Still have to figure out
if the ordered available to promise quantity is correct in this case.

Commit:
Remove unused field

Commit:
Fix test

'partner_id' is not required as the method _get_stock_move_values takes
the 'partner_id' from the 'group'. In fact, it makes the tests fail when
some other non-identified OCA addon is installed.

Commit:
fixup! Rename to stock_available_to_promise_release

Commit:
Improve UI: group moves to release by picking

Commit:
Use sale's confirmation date as priority date

Commit:
Rename field for clarity

Commit:
Rename method for clarity

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -0,0 +1,154 @@
+            and not self.location_id.should_bypass_reservation()
+        )
+
+    def _ordered_available_to_promise(self):
```
Why is this method not part of _compute_ordered_available_to_promise ? Retrieving the value can be done by reading the field.

guewen replied:
As we can't have proper depends, if values change in the transaction, the field will not be recomputed.
We could add invalidations everywhere, not sure that's useful. The field is good for the view, but the code better has to recompute the value on every access. I should add a docstring about this.

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -0,0 +1,154 @@
+            ("product_id", "=", self.product_id.id),
+            ("date_priority", "<=", self.date_priority),
+            ("warehouse_id", "=", self.warehouse_id.id),
+        ]
```
Ideally we should exclude all the moves planned in the future that you can resupply on time. For example, there is no reason a delivery planned 3 months is blocking quantity while you can resupply in 7 days.
Simple solution would be configuring an horizon for considering blocking moves. A more complex and accurate solution would be by looking at the warehouse product orderpoint planned date.
To support sales order commitment dates, something must be done

jgrandguillaume replied:
We go for the horizon solution here

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -0,0 +1,154 @@
+                continue
+            # do not use the computed field, because it will keep
+            # a value in cache that we cannot invalidate declaratively
+            available_quantity = move._ordered_available_to_promise()
```
use the field "move._ordered_available_to_promise" instead

guewen replied:
There is the explanation in the comment just above why the field is not used ;)
I can rephrase if not understandable enough.

Commit:
Exclude canceled moves from computation

guewen commented:
Work continued on 13.0 on #799
